### PR TITLE
[DO NOT MERGE] Add df.ww.to_parquet method to serialize to a single file

### LINKED ISSUE #1223: Add df.ww.to_parquet method to save data and typing info in specified file
As a user, I wish I could use Woodwork to save my Woodwork dataframe as a parquet file with a filename/path that I specify. Currently, Woodwork serialization accepts only a path and not a file name, and stores the data in a subdirectory named `data` and sets the filename based on the name of the dataframe.  I would like to have the ability to specify the path/filename directly and have Woodwork save the data there. 

The filename could be an optional parameter, and Woodwork could name the file based on the datafame name as it does today, if the user does not specify the filename.

Related to #968 

#### Code Example

```python
df.ww.to_parquet(path=my_directory, filename="my_file.parquet")
```

### LINKED ISSUE #1225: Read parquet files created with df.ww.to_parquet
As a user, I wish I could read a Woodwork initialized dataframe from a parquet or arrow file created with the `df.ww.to_parquet` function. The Woodwork typing information should be read in from the metadata in the file header.

Related Issues: #1223 and #1224

### PR DESCRIPTION
**Pull request by thehomebrewnerd**
Add df.ww.to_parquet method to serialize to a single file
- Closes #1223 
- Closes #1225 

Adds a new accessor method to serialize a Woodwork dataframe to the specified parquet file. Saves all Woodwork typing information and other metadata in the parquet file metadata. Also adds a new method `read_parquet` to deserialize the saved information back into a Woodwork initialized dataframe.

### CODE REVIEW

File: woodwork/deserialize.py
Comment (tamargrey) [inline woodwork/deserialize.py:254]: nitpick: could you do `elif _is_s3(path)` for this block and have the `else` be at the same level

File: woodwork/deserialize.py
Comment (tamargrey) [inline woodwork/deserialize.py:269]: Should the string `"woodwork_metadata"` be a constant somewhere that gets used at `serialize` as well?

File: woodwork/deserialize.py
Comment (tamargrey) [inline woodwork/deserialize.py:236]: Would it be worth indicating that it looks for. the woodwork typing info under the `"woodwork_metadata"` string? If someone has serialized and deserialized through woodwork, there's no real reason to have to know this, but it could be useful if there's some further manipulation a user wanted to do.  (Though maybe we specifically would want to avoid that, in which case, we shouldn't overly advertise location where the typing info is stored within the metadata

Comment (thehomebrewnerd) [reply woodwork/deserialize.py:236]: I'm a bit on the fence here, but generally I am leaning toward not advertising this too much, as I don't think the intent would be for the users to do anything with the Woodwork metadata that gets stored. One reason I could see to list this would be to help advanced users avoid accidentally corrupting the WW data if they modify the parquet file metadata.

I'm thinking it might be best to update the key to include an underscore to better indicate that it isn't supposed to be modified by the users. Changes there, if not done properly, could impact the ability to deserialize the file, and isn't something I would expect casual users to do. 

Maybe we can revisit in the future if we find a situation where users need to know about the details, but for now I think I'd leave this out of the documentation. You ok with this?

File: woodwork/deserialize.py
Comment (thehomebrewnerd) [reply woodwork/deserialize.py:269]: Good idea!

File: woodwork/deserialize.py
Comment (thehomebrewnerd) [reply woodwork/deserialize.py:254]: Yeah, good call. I will clean this up.

File: woodwork/deserialize.py
Comment (tamargrey) [reply woodwork/deserialize.py:236]: I like adding an underscore to the key a lot! And maybe a more genera note that users should be careful to not modify the parquet file metadata, as it may impact deserialization.

**Review by tamargrey — APPROVED** [review]
This looks good to me, but probably worth also getting another set of eyes on this before merging.

**Comment by thehomebrewnerd** [comment]
> Why treat this is as a separate serialization method instead of updating the parquet approach of the `to_disk` method?

We *could* do that, but I think that pretty much means a ground-up re-write of `to_disk` and all the other utility functions that are used along the way (both read and write) to handle this type of flow. Currently `to_disk` does not support users specifying the output filename, and I think that is something we want here. Handling parquet files in `to_disk` like they are here would mean that in all the methods and functions we would need unique paths for parquet files because they would work differently than the other formats.

As a bit of a tangent, I'm not sure having a monolithic `to_disk` method is a great approach long term. Because of all the logic branches, it does make it a bit difficult to follow what happens for a given file format, compared to having individual `to_format` methods. I guess that is probably a conversation for a different day though...

File: woodwork/serialize.py
Comment (davesque) [inline woodwork/serialize.py:249]: Something we might want to do here is this:
```suggestion
        table_meta = table.schema.metadata
        table_meta[METADATA_KEY.encode()] = json.dumps(ww_typing).encode()
        table = table.replace_schema_metadata(table_meta)
```
The advantage of doing it this way is that we would always overwrite the `METADATA_KEY` key which I believe is what's intended here.  If that's not the intention, the old approach is also fine I think.

File: woodwork/deserialize.py
Comment (davesque) [inline woodwork/deserialize.py:268]: Any reason we wouldn't just do this here to keep things consistent?
```suggestion
    ww_metadata = file_metadata.metadata.get(METADATA_KEY.encode())
```
Or maybe another thing we could do is just something like this:
```python
METADATA_KEY = b"_woodwork_metadata"
```
Since we always pass it to arrow which requires keys as bytes.

**Review by davesque — COMMENTED** [review]
Mostly looks good.  Just had a couple minor thoughts/questions.

File: woodwork/deserialize.py
Comment (thehomebrewnerd) [reply woodwork/deserialize.py:116]: As I have been going through this update, I have also come to a similar conclusion that our entire serialization/deserialization approach might benefit from some refactoring. Basically, what led me to this conclusion, is that it's really not possible to change the way we serialize to parquet files inside of our `ww.df.to_disk` method, without changing a lot of other helper functions that are also used for serializing to other file types.

To me, it feels like we should be able to change the way we serialize to parquet without potentially impacting/breaking the way we serialize to other formats - possibly by just updating a `ParquetSerializer` class or something like that. The fact that we can't do that right now leads me to believe there might be a better design for our serialization code.

I suggest we give things a bit more thought before we merge in this PR and decide how we want to proceed.

File: woodwork/serialize.py
Comment (thehomebrewnerd) [reply woodwork/serialize.py:249]: Yeah, that's fair and probably a little cleaner. I mostly copied this code from another example on how to update parquet metadata and didn't think through that detail enough.

File: woodwork/deserialize.py
Comment (thehomebrewnerd) [reply woodwork/deserialize.py:268]: No reason, that was just an oversight. I like the idea of updating `METADATA_KEY` to be bites from the start so we can avoid calling `.encode()` in several places.

File: woodwork/deserialize.py
Comment (davesque) [reply woodwork/deserialize.py:116]: Yeah, interesting thoughts.  And actually what you describe is I think a classic use-case for OO programming.  You have a method that does some piece of a process or algorithm and then you want to have another thing like that but just have it do something slightly different.  So you make a subclass of the class that contains the method and override it with your extra behavior.  That could help with the situation you describe where it's hard to change one helper function without impacting a lot of other things.  Having that helper function in a class would allow changing it in a subclass without impacting other things that depend on the base class.

**Comment by thehomebrewnerd** [comment]
Closing this MR as we have decided to refactor the serialization approach in Woodwork. That refactor will have a significant impact on the changes needed to serialize to a single parquet file, and will be better to handle in a new PR after the serialization refactor is complete.

**Pull request closed without merging**
